# 4.8.3. Relación entre los esfuerzos principales obtenidos en el análisis bidimensional y tridimensional (para el caso de deformación plana)

|Quién | Correo | Fecha | Qué hizo |
| ---  | ---    | ---   | ---      |
|Michael Heredia Pérez      | <mherediap@unal.edu.co> | Julio 20, 2022   | Conversión de código de MATLAB a Python     |
|Michael Heredia Pérez      | <mherediap@unal.edu.co> | Octubre 12, 2022 | Completando el código                       |
|Diego Andrés Alvarez Marín | <daalvarez@unal.edu.co> | Febrero 8, 2024  | Comentarios adicionales y mejora del código |

In [1]:
import numpy as np

# imprime todo con 4 cifras decimales
num_decimales = 4
np.set_printoptions(precision=num_decimales)

# función para imprimir los vectores o matrices de numpy bonitos
def imprimir_vect(vec):
    return f"{np.array2string(vec, formatter={'float': lambda x: f'{x: .{num_decimales}f}'}, separator=', ')}"

Asumamos las siguientes cantidades:

In [2]:
nu = 0.30  # coeficiente de Poisson

sx  = -1   # [Pa]
sy  =  2   # [Pa]
txy = -3   # [Pa]

En el caso de deformación plana se tiene adicionalmente que:

In [3]:
ez  = 0
gxz = 0
gyz = 0

por lo que a partir de la ley de Hooke tenemos que:
\begin{align*}
\tau_{xz} &= G\gamma_{xz} = 0 \text{ Pa} \\
\tau_{yz} &= G\gamma_{yz} = 0 \text{ Pa} \\
\sigma_z  &= \nu(\sigma_x + \sigma_y)
\end{align*}

In [4]:
txz = 0
tyz = 0
sz  = nu*(sx + sy)

De donde resulta que:

In [5]:
print(f"σz = {sz: .{num_decimales}f} Pa")

σz =  0.3000 Pa


## Solución del problema de forma bidimensional

Para calcular los esfuerzos y direcciones principales asociados ensamblamos la matriz de esfuerzos
\begin{equation*}
\boldsymbol{\sigma} =
\begin{pmatrix} 
  \sigma_x   & \tau_{xy} \\ 
  \tau_{xy}  & \sigma_y 
\end{pmatrix}
\end{equation*}

In [6]:
sigma = np.array([[  sx, txy ],
                  [ txy,  sy ]])

sigma

array([[-1, -3],
       [-3,  2]])

y le calculamos sus valores y vectores propios

In [7]:
valp, vecp = np.linalg.eigh(sigma)

In [8]:
# imprimimos los valores propios s2xy, s1xy
valp

array([-2.8541,  3.8541])

In [9]:
# imprimimos los vectores propios correspondientes [n2_2D, n1_2D]
vecp

array([[-0.8507, -0.5257],
       [-0.5257,  0.8507]])

In [10]:
# eigh() arroja los valores propios ordenados de menor a mayor (en orden ascendente), por lo que
s2xy,  s1xy  = valp
n2_2D, n1_2D = vecp.T

In [11]:
print(f"σ1_xy = {s1xy: .{num_decimales}f} Pa,    n1_2D = " + imprimir_vect(n1_2D))
print(f"σ2_xy = {s2xy: .{num_decimales}f} Pa,    n2_2D = " + imprimir_vect(n2_2D))

σ1_xy =  3.8541 Pa,    n1_2D = [-0.5257,  0.8507]
σ2_xy = -2.8541 Pa,    n2_2D = [-0.8507, -0.5257]


## Solución del problema de forma tridimensional

Para calcular los esfuerzos y direcciones principales asociados ensamblamos la matriz de esfuerzos
\begin{equation*}
\boldsymbol{\sigma} =
\begin{pmatrix} 
  \sigma_x   & \tau_{xy} & 0 \\ 
  \tau_{xy}  & \sigma_y  & 0 \\
   0         &     0     & \nu(\sigma_x + \sigma_y) 
\end{pmatrix} \label{eq:mattensdefplana}
\end{equation*}

In [12]:
sigma = np.array([[ sx, txy,            0],
                  [txy,  sy,            0],
                  [  0,   0, nu*(sx + sy)]])

sigma

array([[-1. , -3. ,  0. ],
       [-3. ,  2. ,  0. ],
       [ 0. ,  0. ,  0.3]])

y le calculamos sus valores y vectores propios

In [13]:
valp, vecp = np.linalg.eigh(sigma)

In [14]:
# imprimimos los valores propios s3, s2, s1
valp

array([-2.8541,  0.3   ,  3.8541])

In [15]:
# imprimimos los vectores propios correspondientes [n3, n2, n1]
vecp

array([[-0.8507,  0.    , -0.5257],
       [-0.5257,  0.    ,  0.8507],
       [-0.    ,  1.    ,  0.    ]])

In [16]:
# eigh() arroja los valores propios ordenados de menor a mayor (en orden ascendente), por lo que
s3, s2, s1 = valp
n3, n2, n1 = vecp.T

Observe que $\sigma_1 \geq \sigma_2 \geq \sigma_3$:

In [17]:
print(f"σ1 = {s1: .{num_decimales}f} Pa,    n1 = " + imprimir_vect(n1))
print(f"σ2 = {s2: .{num_decimales}f} Pa,    n2 = " + imprimir_vect(n2))
print(f"σ3 = {s3: .{num_decimales}f} Pa,    n3 = " + imprimir_vect(n3))

σ1 =  3.8541 Pa,    n1 = [-0.5257,  0.8507,  0.0000]
σ2 =  0.3000 Pa,    n2 = [ 0.0000,  0.0000,  1.0000]
σ3 = -2.8541 Pa,    n3 = [-0.8507, -0.5257, -0.0000]


Verificamos que los vectores principales formen un sistema coordenado de la mano derecha, a partir del producto cruz $\boldsymbol{\hat{n}}_3 = \boldsymbol{\hat{n}}_1 \times \boldsymbol{\hat{n}}_2$:

In [18]:
n3_calculado = np.cross(n1, n2)

print("n3_calculado = ", imprimir_vect(n3_calculado))

n3_calculado =  [ 0.8507,  0.5257, -0.0000]


Fíjese que obtenemos el vector $\boldsymbol{\hat{n}}_3$ con sentido contrario; esto se puede solucionar considerando en sentido contrario a $\boldsymbol{\hat{n}}_2$:

In [19]:
# Calculo n3 = n1 x -n2
n3_calculado = np.cross(n1, -n2)

print("n3_calculado = ", imprimir_vect(n3_calculado))

n3_calculado =  [-0.8507, -0.5257,  0.0000]


A diferencia del código de Matlab usado en el `main.pdf`, aquí el ajuste de signos es diferente.

En conclusión:

In [20]:
n1_3D =  n1
n2_3D = -n2
n3_3D =  n3

## Análisis de resultados

En el caso bidimensional se tiene que:

In [24]:
print(f"σ1_xy = {s1xy: .{num_decimales}f} Pa,    n1_2D = " + imprimir_vect(n1_2D))
print(f"σ2_xy = {s2xy: .{num_decimales}f} Pa,    n2_2D = " + imprimir_vect(n2_2D))
print(f"σz    = {  sz: .{num_decimales}f} Pa")

σ1_xy =  3.8541 Pa,    n1_2D = [-0.5257,  0.8507]
σ2_xy = -2.8541 Pa,    n2_2D = [-0.8507, -0.5257]
σz    =  0.3000 Pa


Y en el tridimensional que:

In [22]:
print(f"σ1   = {s1: .{num_decimales}f} Pa,    n1_3D = " + imprimir_vect(n1_3D))
print(f"σ2   = {s2: .{num_decimales}f} Pa,    n2_3D = " + imprimir_vect(n2_3D))
print(f"σ3   = {s3: .{num_decimales}f} Pa,    n3_3D = " + imprimir_vect(n3_3D))

σ1   =  3.8541 Pa,    n1_3D = [-0.5257,  0.8507,  0.0000]
σ2   =  0.3000 Pa,    n2_3D = [-0.0000, -0.0000, -1.0000]
σ3   = -2.8541 Pa,    n3_3D = [-0.8507, -0.5257, -0.0000]


Con respecto a los esfuerzos principales se cumple que:
\begin{align*}
\sigma_1 &= \max\left(\left(\sigma_1\right)_{xy},\ \left(\sigma_2\right)_{xy},\ \nu\left(\sigma_x + \sigma_y\right)\right) \\
\sigma_2 &= \operatorname{mediana}\left(\left(\sigma_1\right)_{xy},\ \left(\sigma_2\right)_{xy},\ \nu\left(\sigma_x + \sigma_y\right)\right) \\
\sigma_3 &= \min\left(\left(\sigma_1\right)_{xy},\ \left(\sigma_2\right)_{xy},\ \nu\left(\sigma_x + \sigma_y\right)\right).
\end{align*}

In [23]:
s1 = np.max   ([s1xy, s2xy, nu*(sx + sy)])
s2 = np.median([s1xy, s2xy, nu*(sx + sy)])
s3 = np.min   ([s1xy, s2xy, nu*(sx + sy)])

print(f"σ1 = {s1: .{num_decimales}f} Pa")
print(f"σ2 = {s2: .{num_decimales}f} Pa")
print(f"σ3 = {s3: .{num_decimales}f} Pa")

σ1 =  3.8541 Pa
σ2 =  0.3000 Pa
σ3 = -2.8541 Pa


En resumen, se tiene que dos de los valores propios corresponden a $(\sigma_1)_{xy}$ y $(\sigma_2)_{xy}$, mientras que el otro corresponde a $\sigma_z$. Los vectores propios asociados son los mismos calculados en el caso bidimensional para los dos primeros, agregando un 0 en la componente $z$, mientras que el vector propio asociado a $\sigma_z$ tiene la dirección de $[0,0, \pm 1]^T$ , siendo necesario ajustar su sentido. 

## Cálculo del esfuerzo cortante máximo

El esfuerzo cortante máximo se calcula como:
$$
\tau_{\text{máx}} = \frac{\sigma_1-\sigma_3}{2}
$$
y se presenta en los planos ortogonales a los vectores $\frac{\boldsymbol{\hat{n}}_1 - \boldsymbol{\hat{n}}_3}{\|\boldsymbol{\hat{n}}_1 - \boldsymbol{\hat{n}}_3\|}$ y $\frac{\boldsymbol{\hat{n}}_1 + \boldsymbol{\hat{n}}_3}{\|\boldsymbol{\hat{n}}_1 + \boldsymbol{\hat{n}}_3\|}$.

In [26]:
tmax = (s1 - s3)/2

tmax

3.3541019662496847

In [29]:
(n1 - n3)/np.linalg.norm(n1 - n3)

array([0.2298, 0.9732, 0.    ])

In [30]:
(n1 + n3)/np.linalg.norm(n1 + n3)

array([-0.9732,  0.2298,  0.    ])

por lo tanto, el esfuerzo cortante máximo es 3.3541 Pa y se presenta en los planos con vector normal $\frac{\boldsymbol{\hat{n}}_1 - \boldsymbol{\hat{n}}_3}{\|\boldsymbol{\hat{n}}_1 - \boldsymbol{\hat{n}}_3\|} = [0.2298,\ 0.9732,\ 0]^T$ y $\frac{\boldsymbol{\hat{n}}_1 + \boldsymbol{\hat{n}}_3}{\|\boldsymbol{\hat{n}}_1 + \boldsymbol{\hat{n}}_3\|} = [-0.9732,\ 0.2298,\ 0]^T$.

Tenga en cuenta que el esfuerzo cortante máximo también se podría calcular mediante:
$$
\tau_{\text{máx}} = \max\left(\frac{\left|\left(\sigma_1\right)_{xy} -\nu\left(\sigma_x + \sigma_y\right)\right|}{2},\ \frac{\left|\left(\sigma_2\right)_{xy} - \nu\left(\sigma_x + \sigma_y\right)\right|}{2},\ \frac{\left|\left(\sigma_1\right)_{xy} - \left(\sigma_2\right)_{xy}\right|}{2}\right);
$$
sin embargo, la fórmula de arriba es preferible, ya que es más sencilla.